src/huggingface_hub/file_download.py

```diff
@@ -366,42 +366,87 @@ def _raise_if_offline_mode_is_enabled(msg: Optional[str] = None):
+    2. Handle internal redirection (if `force_internal_redirects=True`).
+    3. Retry in case request fails with a `ConnectTimeout`, with exponential backoff.
+
+    Args:
+        method (`str`):
+            HTTP method, such as 'GET' or 'HEAD'.
+        url (`str`):
+            The URL of the resource to fetch.
+        max_retries (`int`, *optional*, defaults to `0`):
+            Maximum number of retries, defaults to 0 (no retries).
+        base_wait_time (`float`, *optional*, defaults to `0.5`):
+            Duration (in seconds) to wait before retrying the first time.
+            Wait time between retries then grows exponentially, capped by
+            `max_wait_time`.
+        max_wait_time (`float`, *optional*, defaults to `2`):
+            Maximum amount of time between two retries, in seconds.
+        timeout (`float`, *optional*, defaults to `10`):
+            How many seconds to wait for the server to send data before
+            giving up which is passed to `requests.request`.
+        force_internal_redirects (`bool`, *optional*, defaults to `False`)
+            If True, internal (relative) redirection will be resolved even when
+            `allow_redirection` is set to False.
+        **params (`dict`, *optional*):
+            Params to pass to `requests.request`.
     """
+    # 1. Check online mode
     _raise_if_offline_mode_is_enabled(f"Tried to reach {url}")
+
+    # 2. Force internal redirection
+    if force_internal_redirects:
+        response = _request_wrapper(
+            method=method,
+            url=url,
+            max_retries=max_retries,
+            base_wait_time=base_wait_time,
+            max_wait_time=max_wait_time,
+            timeout=timeout,
+            force_internal_redirects=False,
+            **params,
+        )
+
+        # If redirection, we redirect only relative paths.
+        # This is useful in case of a renamed repository.
+        if 300 <= response.status_code <= 399:
+            parsed_target = urlparse(response.headers["Location"])
+            if parsed_target.netloc == "":
+                # This means it is a relative 'location' headers, as allowed by RFC 7231.
+                # (e.g. '/path/to/resource' instead of 'http://domain.tld/path/to/resource')
+                # We want to follow this internal redirect !
+                #
+                # Highly inspired by `resolve_redirects` from requests library.
+                # See https://github.com/psf/requests/blob/main/requests/sessions.py#L159
+                return _request_wrapper(
+                    method=method,
+                    url=urlparse(url)._replace(path=parsed_target.path).geturl(),
+                    max_retries=max_retries,
+                    base_wait_time=base_wait_time,
+                    max_wait_time=max_wait_time,
+                    timeout=timeout,
+                    force_internal_redirects=True,
+                    **params,
+                )
```

Comment:
this will only do *one* redirection (no `while (true)`), but i think moon-landing guarantees that repo A is redirected to repo C directly even if there was another rename, no? cc @Pierrci and also @allendorf who implemented redirects if i remember correctly

Reply:
I reused `force_internal_redirects=True` which will recursively fetch new redirections (so from repo A to repo B to repo C if it occurs). Recursion in Python are not so optimal but I hardly believe we will reach the 1000 chained redirections 😄

Reply:
> i think moon-landing guarantees that repo A is redirected to repo C directly even if there was another rename

yes, there shouldn't be any recursive redirection (it won't hurt if you support them out of the box, they should just never happen)